sql-plugin/src/main/scala/com/nvidia/spark/rapids/NvtxWithMetrics.scala

```diff
@@ -27,31 +29,75 @@ object NvtxWithMetrics {
   }
 }
 
+object ThreadLocalMetrics {
+  val addressOrdering: Ordering[GpuMetric] = Ordering.by(System.identityHashCode(_))
```

Comment:
Would it be simpler to have each GpuMetric track itself? This feels like we are adding in a lot of overhead with the ThreadLocal values.

```
sealed abstract class GpuMetric extends Serializable {
  def value: Long
  def set(v: Long): Unit
  def +=(v: Long): Unit
  def add(v: Long): Unit

  private var isTimerActive = false

  final def tryActivateTimer(): Bool = {
    if (!isTimerActive) {
      isTimerActive = true
      true
    } else {
      // output a warning if this is not NoopMetric
      false
    }
  }

  final def deactivateTimer(duration: Long): Unit = {
    if (isActive) {
      isActive = false
      add(duration)
    }
  }

  final def ns[T](f: => T): T = {
     if (tryActivateTimer()) {
      val start = System.nanoTime()
      try {
        f
      } finally {
        deactivateTimer(System.nanoTime() - start)
      }
    }
  }
}
```

Reply:
I should add that this is mostly about code complexity. I don't think the performance difference will be significant.

Reply:
Hi Bobby, refined the PR accordingly. Threadlocal removed